citeproc-php v2

jonathangreen opened the pull request:
## JIRA Ticket:
https://jira.duraspace.org/browse/ISLANDORA-2243

## What does this Pull Request do?

Replaces the [Islandora fork](https://jira.duraspace.org/browse/ISLANDORA-1015) of [citeproc-php v1](https://github.com/Islandora/citeproc-php) with the new maintained [citeproc v2](https://github.com/seboettg/citeproc-php).

## What's new?

This replaces our fork of citeproc-php with the new maintained version. This solves a number of my issues with citation parsing using the old version of citeproc, it will also lessen the burden of maintaining a fork of the original citeproc code. 

Issues solved using new citeproc-php:
* Improper title capitalization with certain stop words.
* Improper display when a title contains a quoted title.
* WIP https://github.com/seboettg/citeproc-php/issues/50

There are some implications of this PR though: 
* The new citeproc v2 library only supports PHP 5.6+
* The new citeproc v2 library is meant to be installed using composer

The PHP version violates our versioning policy where we support PHP 5.3+ and requiring composer is an unusual step in a Drupal 7 module (although more modules are doing this as well). 

If we want tighter integration with composer we could rely on the [Composer Manager](https://www.drupal.org/project/composer_manager) module, but I didn't want to increase the dependancies further, so I did not do this currently.

## How should this be tested?

* Update module using this patch. 
* Install citeproc php using composer.
  * `composer install`

# Interested parties
@bryjbrown  @Islandora/7-x-1-x-committers

jonathangreen committed:
Update to citeproc v2.

jonathangreen committed:
Fix use of language files.

jonathangreen committed:
Update readme files.

willtp87 commented:
I wonder about if the library usage could be conditional on the php version.

jonathangreen committed:
Coding standards

jonathangreen commented:
@willtp87 / @Islandora/7-x-1-x-committers do you think that a toggle to choose which version of the library to use would be an acceptable solution?

willtp87 commented:
Toggle would be fine. Auto would be gold.

jonathangreen committed:
Coding standards II

jonathangreen commented:
Does auto actually make sense though? You update the version of PHP your using, and suddenly the module stops working because its looking for a different version of the library in a different location. I think I would find that surprising...

willtp87 commented:
@jonathangreen  good point, don't see a way around it for our system.

jonathangreen committed:
Coding standards III

DiegoPino commented:
@jonathangreen you know I agree with composer based installation is Drupal. I even feel it is convenient, when doing this on a module level, that each one has its own vendor (for today's standards that extra space is like nothing) instead of having a central vendor folder that needs to work for everyone. Other than that, I know @willtp87 problem with this (5.6> v/s older PHP) is valid and breaks some jailed-stuck deployments. Updating PHP is really really not an issue, but having a toggle that breaks deployment is not good neither as you state here https://github.com/Islandora/islandora_scholar/pull/291#issuecomment-393513474

So, i have a proposal. Let's make a second islandora citeproc module (submodule) here that triggers a HUGE warning if enabled on 5.5<= and we do the same HUGE warning for current citeproc. We document it and tell people that they need to run either ones? And both can even share settings.

Too much work?

Lastly i would recommend the use of hook_init() as seen here
https://github.com/mnylc/islandora_multi_importer/blob/master/islandora_multi_importer.module#L15-L20
for the the autoload, makes the reusue of whatever you load inside the module and in subfolders, etc, much easier.

bryjbrown commented:
+1 for a toggle instead of auto, toggling would make testing a move from one to the other much easier as well.

jonathangreen committed:
Fix spelling

jonathangreen committed:
Update citeproc to be able to use either library version.

jonathangreen committed:
Update readme files

jonathangreen committed:
Whitespace cleanup

jonathangreen committed:
Fix issues from testing.

jonathangreen committed:
Add end of file returns

jonathangreen commented:
Added some more commits to allow you to choose what version of citeproc you are using. Also added a update hook, so if you are already using citeproc then it will keep you on the old version. For new installs it will choose the version to use based on PHP version until you select one.

jonathangreen committed:
Coping standards IV - Hyper-standards

jonathangreen committed:
Coding Standards V - Return of the doc comment.

jonathangreen commented:
Another advantage to using this implementation is that [OJS](https://pkp.sfu.ca/ojs/) is using this same library [1][2].

[1] https://github.com/seboettg/citeproc-php/issues/43
[2] https://github.com/pkp/pkp-lib/issues/723

willtp87 commented on modules/citeproc/citeproc.module:
```diff
@@ -5,13 +5,48 @@
+function citeproc_init() {
+  module_load_include('inc', 'citeproc', 'includes/utilities');
+  citeproc_default_backend();
+  $backend = variable_get('citeproc_backend', citeproc_default_backend());
```
Dynamic variable defaults have caused me a few headaches in the past.  I suggest either setting the variable in hook_install or using `NULL` to indicate it has not been configured.

jonathangreen committed:
Set default in hook_install

jonathangreen commented:
Addressed code review feedback.

DiegoPino commented on modules/citeproc/citeproc.install:
```diff
@@ -13,3 +13,29 @@ function citeproc_update_7100() {
+ */
+function citeproc_install() {
+  module_load_include('inc', 'citeproc', 'includes/utilities');
+  $default = citeproc_v2_supported() ? CITEPROC_BACKEND_V2 : CITEPROC_BACKEND_V1;
```
Not sure if the constants defined in .module will be present during bootstrap slimmed install hook. We have this issue in openseadragon. Errors only if installed for the first time. Can we live without constants? Also reviewing pulls from Minneapolis airport so i could be missing something! Resto looks good!!

DiegoPino reviewed (changes requested):
Review use of constants in .install

jonathangreen commented:
Is that true Diego? I hadn’t tested but I did read the ‘hook_install’ docs when I wrote it and it specifically says the module file is loaded.

> The implementation can rely on the .module file being loaded. 
https://api.drupal.org/api/drupal/modules%21system%21system.api.php/function/hook_install/7.x

DiegoPino commented:
Will confirm once i can test tonight! If i’m mistaken will totally approve

El El lun, 4 de jun. de 2018 a las 19:32, Jonathan Green <
notifications@github.com> escribió:

> Is that true Diego? I hadn’t tested but I did read the ‘hook_install’ docs
> when I wrote it and it specifically says the module file is loaded.
>
> The implementation can rely on the .module file being loaded.
>
> https://api.drupal.org/api/drupal/modules%21system%21system.api.php/function/hook_install/7.x
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_scholar/pull/291#issuecomment-394542599>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn8563x4mXQuKT5IER9D45vI2Wa5s8yks5t5dGrgaJpZM4UUD6U>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

jonathangreen commented:
Hey @DiegoPino, I just gave this a test, and it seems to work fine for me depending on a constant in the module file. I can disable, uninstall, see the variable is gone. Enable again and check that the variable is correctly set and no warnings are shown.

DiegoPino commented:
Cool. My  second  ✈️ got delayed so got a chance to see this. Will approve
and merge in 24 hours, know i need to find the issue with the open
seadragon install file and that version constant!! Thanks

El El jue, 7 de jun. de 2018 a las 16:27, Jonathan Green <
notifications@github.com> escribió:

> Hey @DiegoPino <https://github.com/DiegoPino>, I just gave this a test,
> and it seems to work fine for me depending on a constant in the module
> file. I can disable, uninstall, see the variable is gone. Enable again and
> check that the variable is correctly set and no warnings are shown.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_scholar/pull/291#issuecomment-395585482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn85zvQkYsJdnNrGdRv3nVWKcfHp4LJks5t6ajEgaJpZM4UUD6U>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

DiegoPino reviewed (approved):
Concerns addresses. Looks good. Will merge in 24 hours. Thanks Jon for investigating my paranoia with CONSTANTS

DiegoPino commented:
👀 @jonathangreen good work on this. Pretty smooth
